x/authenticator/ante/authenticator.go

```diff
@@ -0,0 +1,137 @@
+package ante
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+)
+
+// Verify all signatures for a tx and return an error if any are invalid. Note,
+// the AuthenticatorDecorator will not check signatures on ReCheck.
+//
+// CONTRACT: Pubkeys are set in context for all signers before this decorator runs
+// CONTRACT: Tx must implement SigVerifiableTx interface
+type AuthenticatorDecorator struct {
+	ak              authante.AccountKeeper
+	signModeHandler authsigning.SignModeHandler
+}
+
+func NewAuthenticatorDecorator(
+	ak authante.AccountKeeper,
+	signModeHandler authsigning.SignModeHandler,
+) AuthenticatorDecorator {
+	return AuthenticatorDecorator{
+		ak:              ak,
+		signModeHandler: signModeHandler,
+		//TODO: Add authenticator struct here for later user
+	}
+}
+
+// AnteHandle is the authenticator decorator ante handler
+// this is used to validate multiple signatures
+func (ad AuthenticatorDecorator) AnteHandle(
+	ctx sdk.Context,
+	tx sdk.Tx,
+	simulate bool,
+	next sdk.AnteHandler,
+) (newCtx sdk.Context, err error) {
+	ad.ValidateSignature(ctx, tx, simulate)
+	if err != nil {
+		return ctx, err
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// ValidateSignature validates the txn signature using Secp256k1 keys
+func (ad AuthenticatorDecorator) ValidateSignature(
+	ctx sdk.Context,
+	tx sdk.Tx,
+	simulate bool,
+) (newCtx sdk.Context, err error) {
+	sigTx, ok := tx.(authsigning.SigVerifiableTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "invalid transaction type")
+	}
+
+	// stdSigs contains the sequence number, account number, and signatures.
+	// When simulating, this would just be a 0-length slice.
+	sigs, err := sigTx.GetSignaturesV2()
+	if err != nil {
+		return ctx, err
+	}
+
+	signerAddrs := sigTx.GetSigners()
+
+	// check that signer length and signature length are the same
+	if len(sigs) != len(signerAddrs) {
+		return ctx, sdkerrors.Wrapf(
+			sdkerrors.ErrUnauthorized,
+			"invalid number of signer;  expected: %d, got %d",
+			len(signerAddrs),
+			len(sigs))
+	}
+
+	for i, sig := range sigs {
```

Comment:
we should probably abstract the inside of this loop too. I'm thinking that code block is what the `ClassicAuthorizer` will call